pkg/controller/framework/shared_informer.go

```diff
@@ -279,21 +279,21 @@ func (p *processorListener) add(notification interface{}) {
 func (p *processorListener) pop(stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 
-	p.lock.Lock()
-	defer p.lock.Unlock()
 	for {
+		p.lock.Lock()
```

Comment:
Ideally, I prefer opaquing blocks that should be guarded with lock into
exit = func() bool {
  p.lock.Lock()
  defer p.lock.Unlock()
  ...
}()
if exit {
  return
}
// now the rest of the thing that shouldn't be under lock


The above pattern is better, because if you add return in a different place and forget to unlock, this will result in deadlock.